sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -287,76 +289,145 @@ private SpecializedBlobClientBuilder prepareBuilder() {
+
+                final String blockId = Base64.getEncoder().encodeToString(
+                    UUID.randomUUID().toString().getBytes(UTF_8));
+
+                return blockBlobAsyncClient.stageBlockWithResponse(blockId, progressData, buffer.remaining(),
+                    null, accessConditions.getLeaseId())
+                    // We only care about the stageBlock insofar as it was successful,
+                    // but we need to collect the ids.
+                    .map(x -> blockId)
+                    .doFinally(x -> pool.returnBuffer(buffer))
+                    .flux();
+            }) // TODO: parallelism?
+            .collect(Collectors.toList())
+            .flatMap(ids ->
+                blockBlobAsyncClient.commitBlockListWithResponse(ids, headers, metadata, tier, accessConditions));
+    }
+
+    private Mono<Response<BlockBlobItem>> determineUploadFullOrChunked(final Flux<ByteBuffer> data,
+        final Function<Flux<ByteBuffer>, Mono<Response<BlockBlobItem>>> uploadInChunks,
+        final BiFunction<Flux<ByteBuffer>, Long, Mono<Response<BlockBlobItem>>> uploadFullBlob) {
+        final long chunkedUploadRequirement = 4 * Constants.MB;
+        final long[] bufferedDataSize = {0};
+        final LinkedList<ByteBuffer> cachedBuffers = new LinkedList<>();
+
+        /*
+         * Window the reactive stream until either the stream completes or the windowing size is hit. If the window
+         * size is hit the next emission will be the pointer to the rest of the reactive steam.
+         *
+         * Once the windowing has completed buffer the two streams, this should create a maximum overhead of ~4MB plus
+         * the next stream emission if the window size was hit. If there are two streams buffered use Stage Blocks and
+         * Put Block List as the upload mechanism otherwise use Put Blob.
+         */
+        return data
+            .windowUntil(buffer -> {
+                if (bufferedDataSize[0] > chunkedUploadRequirement) {
+                    return false;
+                } else {
+                    bufferedDataSize[0] += buffer.remaining();
+
+                    if (bufferedDataSize[0] > chunkedUploadRequirement) {
+                        return true;
+                    } else {
+                        /*
+                         * Buffer until the first 4MB are emitted from the stream in case Put Blob is used. This is
+                         * done to support replayability which is required by the Put Blob code path, it doesn't buffer
+                         * the stream in a way that the Stage Blocks and Put Block List code path does, and this API
+                         * explicitly states that it supports non-replayable streams.
+                         */
+                        ByteBuffer cachedBuffer = ByteBuffer.allocate(buffer.remaining()).put(buffer);
+                        cachedBuffer.position(0);
```

Comment:
Consider using `cachedBuffer.flip()` instead. It updates the state of a few things underneath -  like limit, mark and position which may be relevant to something downstream.

Reply:
Will use flip